### ISSUE TITLE
com.google.gson.internal.LinkedTreeMap cannot be cast to custom Parceler object

### ISSUE THREAD

**Issue by ardoramor** [johncarl81/parceler#85]
Hi John, ran into an error for the first time with using the ParcelWrapperConverter, which you kindly provided for issue #74.

I am getting the following exception:

`java.lang.ClassCastException: com.google.gson.internal.LinkedTreeMap cannot be cast to com.test.api.model.ServerConfig`

Server returns the following JSON via Retrofit:

    {
        "code":"OK",
        "message":"Successfully retrieved system information",
        "data":{
            "anonymousaccountcreation":true,
            "anonymousaccess":true,
            "devicetokenexpiration":2592000
        },
        "securityclassification":"UNCLASSIFIED"
    }

*BaseResponse.java*

	@Parcel
	public class BaseResponse<T> {
		private static final String TAG = LogUtils.produceTag(BaseResponse.class);

		private static final String RESPONSE_CODE_SUCCESS = "OK";

		public static final int STATUS_ERROR = 0;
		public static final int STATUS_OK = 1;

		public String code;
		public String message;
		@SerializedName("securityclassification")
		public String classification;
		@ParcelPropertyConverter(ParcelWrapperConverter.class)
		public Object data;

		public T getData() {
			return (T)data; // This throws the exception
		}

		public int getStatus() {
			if (code != null && RESPONSE_CODE_SUCCESS.equals(code.toUpperCase())) {
				return STATUS_OK;
			} else {
				return STATUS_ERROR;
			}
		}

		public String getErrorMessage() {
			return message;
		}

		public static class ParcelWrapperConverter implements ParcelConverter {

			@Override
			public void toParcel(Object input, android.os.Parcel parcel) {
				parcel.writeParcelable(Parcels.wrap(input), 0);
			}

			@Override
			public Object fromParcel(android.os.Parcel parcel) {
				return ((ParcelWrapper)parcel.readParcelable(ParcelWrapperConverter.class.getClassLoader())).getParcel();
			}
		}
	}

*ServerConfig.java*

	@Parcel
	public class ServerConfig {
		@SerializedName("anonymousaccountcreation")
		public boolean anonymousAcctCreation;
		@SerializedName("anonymousaccess")
		public boolean anonymousAccess;
		@SerializedName("devicetokenexpiration")
		public int deviceTokenExpiration;
	}

Retrofit interface definition:

    @GET("/api/system/info")
    Observable<BaseResponse<ServerConfig>> getServerConfig();

Method that invokes Retrofit API:

    public Observable<ServerConfig> getServerConfig() {
        return mApi.getServerConfig()
                .doOnNext(new ValidateResponse())
                .map(new Func1<BaseResponse<ServerConfig>, ServerConfig>() {
                    @Override
                    public ServerConfig call(BaseResponse<ServerConfig> response) {
                        return response.getData(); // Exception is thrown in this method
                    }
                });
     }

So basically, when I call `getServerConfig()`, the `response.getData()` attempts to cast `BaseResponse.data` as `ServerConfig`, which apparently conflicts with the way the JSON was interpreted (as `LinkedTreeMap`).

I have tested `BaseResponse` where data was a `Map` before:

    {
        "code":"OK",
        "message":"Some message",
        "data":{
            "a":{ ... },
            "b":{ ... }
        },
        "securityclassification":"UNCLASSIFIED"
    }

And the interface would cast it as `Map<String, CustomParcelerClass>` and it worked. Seems like `ParcelWrapperConverter` is unable to guess at the custom Parceler class. Is there any way to hint at it. Since `BaseResponse` is aware of the `class` of the data, can we use `T` to hint `ParcelWrapperConverter` through annotations?

**Comment by johncarl81** [comment]
Gson is probably not marshaling into your given type since you're using `Object` as the type.  I believe I fixed this so you can use `T` instead in version 0.2.17-SNAPSHOT.  Does this work?:

```java
@Parcel
public class BaseResponse<T> {
    ...
    @ParcelPropertyConverter(ParcelWrapperConverter.class)
    public T data;

    public T getData() {
        return data;
    }
    ...

    public static class ParcelWrapperConverter implements ParcelConverter {

        @Override
        public void toParcel(Object input, android.os.Parcel parcel) {
            parcel.writeParcelable(Parcels.wrap(input), 0);
        }

        @Override
        public Object fromParcel(android.os.Parcel parcel) {
            return ((ParcelWrapper)parcel.readParcelable(ParcelWrapperConverter.class.getClassLoader())).getParcel();
        }
    }
}
```

**Comment by ardoramor** [comment]
*Palm in face* I think it somehow set with me that it still wasn't possible and since it worked with the above come before I didn't change it.

Just tested it and it works. Very strong feature. Allows me to do various validations and not just in the case of the immediate response. Thanks again. Since this isn't an issue, I'll close it.

**Comment by johncarl81** [comment]
Cool :+1:

**Comment by johncarl81** [comment]
Right.  This is not one of the supported types.  You'll have to make a
custom serializer for it.
